# MetaPhlAn

> [MetaPhlAn](https://huttenhower.sph.harvard.edu/metaphlan) is a computational tool for profiling the composition of microbial communities (Bacteria, Archaea and Eukaryotes) from metagenomic shotgun sequencing data (i.e. not 16S) with species-level.

## Profile Format

Taxpasta expects a four-column, tab-separated file. This is generated by redirecting the MetaPhlAn3 `stdout`. It interprets the columns as:

| Column Header      | Description |
| ------------------ | ----------- |
| clade_name         |             |
| taxonomy_id        |             |
| relative_abundance |             |
| additional_species |             |

## Example

```text
#mpa_v30_CHOCOPhlAn_201901
#/workspace/nextflow-mentoring/second_pipeline/work/conda/env-e9a987b33d7e880200ff2190a27ad05d/bin/metaphlan ERR525939.trimmed.R1.fastq.gz,ERR525939.trimmed.R2.fastq.gz --bowtie2out metagenome.bowtie2.bz2 --input_type fastq -o ERR525939.taxa.txt
#SampleID	Metaphlan_Analysis
#clade_name	NCBI_tax_id	relative_abundance	additional_species
k__Bacteria	2	100.0
k__Bacteria|p__Bacteroidetes	2|976	65.30411
k__Bacteria|p__Proteobacteria	2|1224	27.81459
k__Bacteria|p__Firmicutes	2|1239	4.35836
```

## Example

```text
# git tag version 3.0.3 |  motus version 3.0.3 | map_tax 3.0.3 | gene database: nr3.0.3 | calc_mgc 3.0.3 -y insert.scaled_counts -l 75 | calc_motu 3.0.3 -k mOTU -C no_CAMI -g 3 -c -p | taxonomy: ref_mOTU_3.0.3 meta_mOTU_3.0.3
# call: python /usr/local/bin/../share/motus-3.0.1//motus profile -p -c -f ERX5474932_ERR5766176_1.fastq.gz -r ERX5474932_ERR5766176_2.fastq.gz -db db_mOTU -t 2 -n 2612_pe-ERR5766176-db_mOTU -o 2612_pe-ERR5766176-db_mOTU.out
#consensus_taxonomy	NCBI_tax_id	2612_pe-ERR5766176-db_mOTU
Leptospira alexanderi [ref_mOTU_v3_00001]	100053	0
Leptospira weilii [ref_mOTU_v3_00002]	28184	0
Chryseobacterium sp. [ref_mOTU_v3_00004]	NA	0
```
